crates/pipeline_manager/src/auth.rs

```diff
@@ -490,10 +700,27 @@ mod test {
         let modified_base64_claim =
             base64::engine::general_purpose::STANDARD_NO_PAD.encode(claim_str_modified);
         let new_token = token.replace(base64_parts.get(1).unwrap(), modified_base64_claim.as_str());
-        let res = run_test(new_token, validation).await;
-        assert!(matches!(
-            res.err().unwrap(),
-            AuthError::JwtDecoding(e) if *e.kind() == jsonwebtoken::errors::ErrorKind::InvalidSignature
-        ));
+        let req = test::TestRequest::get()
+            .uri("/")
+            .insert_header((http::header::AUTHORIZATION, format!("Bearer {}", new_token)))
+            .to_request();
+        let res = run_test(req, Some(decoding_key), None, validation).await;
+        assert_eq!(401, res.status());
+        assert_eq!(
+            "Bearer error_description=\"InvalidSignature\"",
+            res.headers().get("www-authenticate").unwrap()
+        );
+    }
+
+    #[tokio::test]
+    async fn valid_api_key() {
+        let validation = validation("some-client", "some-iss");
+        let api_key = auth::generate_api_key();
+        let req = test::TestRequest::get()
+            .uri("/")
+            .insert_header((HeaderName::from_str("x-api-key").unwrap(), api_key.clone()))
+            .to_request();
+        let res = run_test(req, None, Some(api_key), validation).await;
+        assert_eq!(200, res.status());
     }
```

Comment:
It would be great to have the following proptest that checks the isolation property:

e.g., client A, and client B: client A only stores things (name, desc, program) with character 'a'; Client B only stores things with character B, all of them invoke random API calls, with either bearer or token, + a client C that has no valid token or API key...

the isolation property is: client a only sees things with 'a', client B with 'b' and client C never sees anything (only gets errors)... wdyt?

Reply:
Nice! These tests aren't yet end-to-end and don't use pipeline manager APIs. But your suggestion sounds like a great way to structure such end-to-end tests. I can work on that after your API hardening, where I can add user info and scopes to the schema.